### DISCUSSION TITLE
Languages faster than Python

### DISCUSSION THREAD

**Discussion by ghost** [community/community#86061] [Programming Help] [answered]
### Body

I am making an MS DOS emulator and I was wondering if you could tell me what is a faster language to use, and what are the cons and pros of using the language you provided?

Also, languages I have some experience in HTML, CSS, JavaScript, NASM X86 for Linux and Python.

In Assembly, I know jumping, creating sections, moving data around, output, and adding.

In JavaScript, I know DOM, if statements, loops, functions, and recursion.

In Python, I know I/O, File manipulation, recursion, if statements, else if statements, else statements and loops.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by nabilhayek** [comment]
Depending on your question whether It's faster as in computation & efficiency or faster as in development time.

1. **Assembly Language (NASM x86):**
   - **Pros:**
     - Extremely low-level, providing direct control over hardware.
     - Can be highly efficient in terms of performance.
     - Good for systems programming and interacting with hardware.
   - **Cons:**
     - Steeper learning curve, especially for complex projects like an emulator.
     - Code may be less readable and maintainable compared to higher-level languages.

2. **C or C++:**
   - **Pros:**
     - Offers low-level control similar to assembly but with more abstraction.
     - Good balance between performance and readability.
     - Widely used for system-level programming and emulation.
   - **Cons:**
     - Learning curve, though less steep than assembly.
     - More verbose compared to higher-level languages.

3. **JavaScript:**
   - **Pros:**
     - Cross-platform, making it easier to share the emulator online.
     - Well-suited for asynchronous operations and event-driven tasks.
     - Your existing knowledge of JavaScript can be leveraged.
   - **Cons:**
     - May not be as performant as lower-level languages for system emulation.
     - Limited direct hardware control compared to assembly or C/C++.

4. **Python:**
   - **Pros:**
     - Easy to learn and read, promoting faster development.
     - Rich ecosystem and libraries can simplify certain tasks.
     - Good for prototyping and quick development.
   - **Cons:**
     - Generally slower than languages like C/C++ due to higher-level abstractions.
     - Limited low-level control over hardware.

A combination of languages might be beneficial. For example, you could use C/C++ for the core emulation engine due to their performance, and then create a user interface using JavaScript and HTML for better cross-platform accessibility.

Ultimately, the choice depends on your priorities — whether it's speed, ease of development, or a balance between the two.

**Comment by thuan2172001** [comment]
Considering your experience and the complexity of the task, Assembly might be a suitable choice if you're comfortable with low-level programming. However, if you want to use another language, you might consider C or C++.

For C, it is more readable and writable for most programmers. It also allows you to perform operations at the hardware level and can be highly optimized. The foreseeable consequence of this language is that C has a learning curve, especially when dealing with pointers and manual memory management.

**Reply by ghost** [reply]
Okay, I think I will use a combination of C++, Python and Assembly. I will use Assembly for the virtual hardware and I’ll use C++ for the emulator just so I have a little more simplicity, but I haven’t thought about what I’ll use Python for.

Thanks for the help!

**Comment by kmadaminjanova0508** [comment, marked as answer]
I think, given the complexity and performance requirements of building an MS DOS emulator, using Assembly (NASM X86) is a traditional and performance-oriented choice. However, considering the complexity and your experience level, you might find it more practical to use a combination of languages. For instance, writing the core emulation engine in Assembly for performance-critical sections and using a higher-level language like Python or JavaScript for the overall emulator framework and user interface. For example, I consider using JavaScript and C++

**Reply by ghost** [reply]
Never mind I think I am too retarded to use C++ again because I have an IQ < 70

**Reply by kmadaminjanova0508** [reply]
If you're finding C++ challenging, it doesn't mean you're "retarded" or incapable. Learning programming languages, especially those like C++, can be complex for many people including myself , I also learned C++ in 6 month still tring to scroll somehow :(

**Comment by ghost** [comment]
What language will be suitable for me?

**Reply by kmadaminjanova0508** [reply]
I think, based on my experience, using C or C++ would be a suitable choice for creating an MS-DOS emulator. These languages offer a good balance between performance and control, and there are existing resources and libraries available to aid in your development. I think  if you're up for the challenge and looking for optimal performance, C/C++ would be a good fit. :)

**Reply by ghost** [reply]
Okay, I will try again.
